File: files/problems/js/rotate-image.js
// by @codeAbinash
// Time Complexity : O(n^2)
// Space Complexity : O(1)

var rotate = function (matrix) {
    let rowLen = matrix[0].length
    let colLen = matrix.length
    for (let i = 0; i < colLen - 1; i++)
        for (let j = i + 1; j < rowLen; j++)
            [matrix[i][j], matrix[j][i]] = [matrix[j][i], matrix[i][j]]
    for (row of matrix) row.reverse()
};

File: leetcode/problems/java/rotate-image.java
class Solution {
  public void rotate(int[][] matrix) {
    int r = matrix[0].length;
    int c = matrix.length;

    for (int i = 0; i < c; i++) {
      for (int j = i; j < r; j++) {
        int temp = matrix[i][j];
        matrix[i][j] = matrix[j][i];
        matrix[j][i] = temp;
      }
    }

    for (int i = 0; i < c; i++) {
      for (int j = 0; j < r / 2; j++) {
        int temp = matrix[i][j];
        matrix[i][j] = matrix[i][r - j - 1];
        matrix[i][r - j - 1] = temp;
      }
    }

  }
}